instructure/canvas-android
# [Student][MBL-12229] File upload persistence

Save file uploads to our database to persist for the new submission flow.
Also adds a progress body for file uploads, updates sent through eventbus ProgressEvent so that anything can receive events.

To test:
This is a pain to actually test. It's not hooked up to any screens yet, so you have to go manually change some files to be able to test it directly.

First - change UploadFilesDialog.kt at line 457 (bottom of `uploadFiles` function) to look like this:
```
            if(bundle != null) {
                dialogCallback?.invoke(EVENT_ON_UPLOAD_BEGIN)
                dialogAttachmentCallback?.invoke(EVENT_ON_UPLOAD_BEGIN, fileList.first())
//                intent.putExtras(bundle)
//                activity?.startService(intent)
                dismiss()
            }
```
Second - change AddSubmissionFragment.kt at line 281 (top of `setupListeners` function) to look like this:
```
        fileUpload.setOnClickListener {
            // TODO: Test new submission service here
            UploadFilesDialog.show(fragmentManager, UploadFilesDialog.createAssignmentBundle(null, canvasContext as Course, assignment)) { _, file ->
//                context?.let { SubmissionService.startFileSubmission(it, canvasContext, assignment.id, assignment.name, assignment.groupCategoryId, arrayListOf(file!!, file, file, file, file, file)) } // to test file submissions, can repeat `file` as many times as desired
//                context?.let { SubmissionService.retryFileSubmission(it, canvasContext, 29) } // to test retries, you need the submission id from the db (set a breakpoint in the submission service and use it)
            }
        }
```

In AddSubmissionFragment, you can uncomment whichever line you want to use for the current thing your testing. You can pull the db files from the device and view them in a db browser, or you can set a breakpoint at the bottom of the SubmissionFileUploadReceiver in SubmissionService.kt to make calls to view the database like:
`db.fileSubmissionQueries.getFilesForSubmissionId(submissionId).executeAsList()`

You can use that to verify that files were input correctly, deleted correctly, set to errors correctly, etc...

Some of the test cases that would be good to hit:
* Submitting one large file or a file over a slower network (to be able to see the progress update in the notification)
* Submitting multiple files
* Submitting multiple files that have an error (the only error I could get is File Quota Exceeded, I can help with getting that test case ready; errors for wrong file type should be caught before submitting the files to the service; the file upload service also stops uploading files as soon as one errors)
* Retrying a submission with some files that uploaded
